### PR TITLE
[Magiclysm] Return of the Elves

### PR DESCRIPTION
**Pull request by Theawesomeboophis**
#### Summary
Mods "Resubmission of #47984 with fixes and a roof"

#### Purpose of change

It's been a project for me for a while and I wanted to finish it

#### Describe the solution

Adds a roof to the workshop and finishes the PR

#### Describe alternatives you've considered

Going to sleep, launching fireworks into the sky, eating a pizza

#### Testing

Tested everything in the last PR, barring the roof.

#### Additional Context

@chaosvolt helped with bugfixing and making the roof because I didn't know what I was doing.

### CODE REVIEW

**Comment by Procyonae** [comment]
What are you trying to do with place_loot? You've used symbols suggesting you want the items to spawn on those symbols but then also used coordinates?

**Comment by Procyonae** [comment]
Without digging it looks like elves is a faction too so you might be able to make a zone over the building to claim it for the elves like the forge of wonders does so they'll defend their stuff too if you wanted

**Comment by Theawesomeboophis** [comment]
> What are you trying to do with place_loot? You've used symbols suggesting you want the items to spawn on those symbols but then also used coordinates?

I'm trying to make the items spawn at those coordinates.

**Comment by Theawesomeboophis** [comment]
> Without digging it looks like elves is a faction too so you might be able to make a zone over the building to claim it for the elves like the forge of wonders does so they'll defend their stuff too if you wanted

The Elves aren't supposed to be very aggressive or defensive, they're supposed to be mostly neutral unless directly attacked, they're not really bothered by you stealing their stuff, they just wanna be helpful in my mind at least.

File: data/mods/Magiclysm/worldgen/elf_workshop.json [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      ],
+      "w": [
+        { "item": "misc_repairkit", "x": [ 0, 23 ], "y": [ 0, 14 ], "chance": 25 },
+        { "item": "hammer", "x": [ 0, 23 ], "y": [ 0, 17 ], "chance": 90 },
+        { "item": "sandpaper", "x": [ 0, 3 ], "y": [ 0, 3 ], "chance": 90 },
+        { "item": "clamp", "x": [ 0, 23 ], "y": [ 0, 23 ], "chance": 90, "repeat": [ 2, 4 ] },
+        { "item": "xacto", "x": [ 0, 23 ], "y": [ 0, 13 ], "chance": 10 }
+      ],
+      "f": [ { "item": "alumentum", "x": [ 0, 23 ], "y": [ 0, 21 ], "chance": 25 } ]
+    }
```
Comment (Procyonae) [inline data/mods/Magiclysm/worldgen/elf_workshop.json:41]:
```suggestion
    "place_loot": [
        { "group": "tools_anvil", "x": [ 0, 24 ], "y": [ 0, 20 ], "chance": 75, "repeat": [ 1, 2 ] },
        { "item": "mithril_ingot", "x": [ 0, 23 ], "y": [ 0, 25 ], "chance": 10 },
        { "item": "scrap_copper", "x": [ 0, 22 ], "y": [ 0, 21 ], "chance": 40 },
        { "group": "tools_anvil", "x": [ 0, 13 ], "y": [ 0, 22 ], "chance": 25 },
        { "item": "clamp", "x": [ 0, 23 ], "y": [ 0, 23 ], "chance": 50, "repeat": [ 1, 3 ] },
        { "item": "metal_file", "x": [ 0, 23 ], "y": [ 0, 15 ], "chance": 50 },
        { "item": "pin_reamer", "x": [ 0, 23 ], "y": [ 0, 21 ], "chance": 25 },
        { "item": "mithril_ingot", "x": [ 0, 15 ], "y": [ 0, 21 ], "chance": 10 },
        { "item": "scrap_copper", "x": [ 0, 23 ], "y": [ 0, 12 ], "chance": 80, "repeat": [ 1, 3 ] },
        { "item": "misc_repairkit", "x": [ 0, 23 ], "y": [ 0, 14 ], "chance": 25 },
        { "item": "hammer", "x": [ 0, 23 ], "y": [ 0, 17 ], "chance": 90 },
        { "item": "sandpaper", "x": [ 0, 3 ], "y": [ 0, 3 ], "chance": 90 },
        { "item": "clamp", "x": [ 0, 23 ], "y": [ 0, 23 ], "chance": 90, "repeat": [ 2, 4 ] },
        { "item": "xacto", "x": [ 0, 23 ], "y": [ 0, 13 ], "chance": 10 },
        { "item": "alumentum", "x": [ 0, 23 ], "y": [ 0, 21 ], "chance": 25 }
    ]
```
I think this is what you want but I think you need to move it out of the palette and into the "object" of the map too bc it's coordinate based

File: data/mods/Magiclysm/worldgen/elf_workshop.json [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  {
+    "type": "overmap_terrain",
+    "id": [ "elf_workshop", "elf_workshop_roof" ],
+    "name": "elf workshop",
+    "sym": "W",
+    "color": "red",
+    "see_cost": 6,
+    "flags": [ "RISK_HIGH" ]
+  }
+]
```
Comment (Procyonae) [inline data/mods/Magiclysm/worldgen/elf_workshop.json:94]:
```suggestion
[
  {
    "type": "mapgen",
    "method": "json",
    "om_terrain": "elf_workshop",
    "object": {
      "fill_ter": "t_metal_floor",
      "rows": [
        "111111111111111111111111",
        "1fa.....fa..........jsw1",
        "1ms......m....d....r.ct1",
        "1d...................ct1",
        "1bsw.........c........d1",
        "1dsw........ctc......af1",
        "1............c.........3",
        "1msa...................1",
        "1d.f..................d1",
        "1......................3",
        "1bsw..................b1",
        "1....d............j....1",
        "1...ws............sw...1",
        "1......................1",
        "1......................3",
        "1......................1",
        "1...ws............sw...1",
        "1......................1",
        "1...ws............sw...1",
        "1......................1",
        "1...........m..........1",
        "1f.........sa.........f1",
        "1a.sm......f.......ms.a1",
        "111111111111111111111111"
      ],
      "terrain": { "3": "t_rdoor_c", "1": "t_strconc_wall", ".": "t_metal_floor" },
      "furniture": {
        "a": "f_anvil",
        "t": "f_table",
        "b": "f_tablesaw",
        "c": "f_chair",
        "d": "f_drill_press",
        "j": "f_jointer",
        "r": "f_router",
        "w": "f_workbench",
        "m": "f_workbench",
        "s": "f_stool",
        "f": "f_demon_forge"
      },
      "place_monsters": [ { "monster": "GROUP_WORKSHOP", "x": [ 3, 5 ], "y": [ 4, 23 ], "density": 0.5, "repeat": [ 1, 6 ] } ],
      "place_loot": [
        { "group": "tools_anvil", "x": [ 0, 24 ], "y": [ 0, 20 ], "chance": 75, "repeat": [ 1, 2 ] },
        { "item": "mithril_ingot", "x": [ 0, 23 ], "y": [ 0, 25 ], "chance": 10 },
        { "item": "scrap_copper", "x": [ 0, 22 ], "y": [ 0, 21 ], "chance": 40 },
        { "group": "tools_anvil", "x": [ 0, 13 ], "y": [ 0, 22 ], "chance": 25 },
        { "item": "clamp", "x": [ 0, 23 ], "y": [ 0, 23 ], "chance": 50, "repeat": [ 1, 3 ] },
        { "item": "metal_file", "x": [ 0, 23 ], "y": [ 0, 15 ], "chance": 50 },
        { "item": "pin_reamer", "x": [ 0, 23 ], "y": [ 0, 21 ], "chance": 25 },
        { "item": "mithril_ingot", "x": [ 0, 15 ], "y": [ 0, 21 ], "chance": 10 },
        { "item": "scrap_copper", "x": [ 0, 23 ], "y": [ 0, 12 ], "chance": 80, "repeat": [ 1, 3 ] },
        { "item": "misc_repairkit", "x": [ 0, 23 ], "y": [ 0, 14 ], "chance": 25 },
        { "item": "hammer", "x": [ 0, 23 ], "y": [ 0, 17 ], "chance": 90 },
        { "item": "sandpaper", "x": [ 0, 3 ], "y": [ 0, 3 ], "chance": 90 },
        { "item": "clamp", "x": [ 0, 23 ], "y": [ 0, 23 ], "chance": 90, "repeat": [ 2, 4 ] },
        { "item": "xacto", "x": [ 0, 23 ], "y": [ 0, 13 ], "chance": 10 },
        { "item": "alumentum", "x": [ 0, 23 ], "y": [ 0, 21 ], "chance": 25 }
      ]
    }
  },
  {
    "type": "overmap_special",
    "id": "elf_workshop",
    "overmaps": [
      { "point": [ 0, 0, 0 ], "overmap": "elf_workshop_north" },
      { "point": [ 0, 0, 1 ], "overmap": "elf_workshop_roof_north" }
    ],
    "locations": [ "wilderness" ],
    "city_distance": [ 20, -5 ],
    "city_sizes": [ 0, 13 ],
    "occurrences": [ 0, 2 ]
  },
  {
    "type": "overmap_terrain",
    "id": [ "elf_workshop", "elf_workshop_roof" ],
    "name": "elf workshop",
    "sym": "W",
    "color": "red",
    "see_cost": 6,
    "flags": [ "RISK_HIGH" ]
  }
]

```
You seemed to miss me saying it needed moving, I just removed the palette seeing as you're only using it for the one map anyways

Comment (Theawesomeboophis) [reply data/mods/Magiclysm/worldgen/elf_workshop.json:94]: Sorry, I was doing something else, I didn't really know what you meant by object: i thought object was a field I didn't know about and I tried looking for it in the files for reference.

Comment (Procyonae) [reply data/mods/Magiclysm/worldgen/elf_workshop.json]: It's cool it's just the one that most stuff goes in below your "om_terrain" id

File: data/mods/Magiclysm/worldgen/elf_workshop.json [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        "j": "f_jointer",
+        "r": "f_router",
+        "w": "f_workbench",
+        "m": "f_workbench",
+        "s": "f_stool",
+        "f": "f_demon_forge"
+      },
+      "place_monsters": [ { "monster": "GROUP_WORKSHOP", "x": [ 3, 5 ], "y": [ 4, 23 ], "density": 0.5, "repeat": [ 1, 6 ] } ],
+      "place_loot": [
+        { "group": "tools_anvil", "x": [ 0, 24 ], "y": [ 0, 20 ], "chance": 75, "repeat": [ 1, 2 ] },
```
Comment (Procyonae) [inline data/mods/Magiclysm/worldgen/elf_workshop.json:50]:
```suggestion
        { "group": "tools_anvil", "x": [ 0, 23 ], "y": [ 0, 20 ], "chance": 75, "repeat": [ 1, 2 ] },
```

File: data/mods/Magiclysm/worldgen/elf_workshop.json [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        "r": "f_router",
+        "w": "f_workbench",
+        "m": "f_workbench",
+        "s": "f_stool",
+        "f": "f_demon_forge"
+      },
+      "place_monsters": [ { "monster": "GROUP_WORKSHOP", "x": [ 3, 5 ], "y": [ 4, 23 ], "density": 0.5, "repeat": [ 1, 6 ] } ],
+      "place_loot": [
+        { "group": "tools_anvil", "x": [ 0, 23 ], "y": [ 0, 20 ], "chance": 75, "repeat": [ 1, 2 ] },
+        { "item": "mithril_ingot", "x": [ 0, 23 ], "y": [ 0, 25 ], "chance": 10 },
```
Comment (Procyonae) [inline data/mods/Magiclysm/worldgen/elf_workshop.json:51]:
```suggestion
        { "item": "mithril_ingot", "x": [ 0, 23 ], "y": [ 0, 23 ], "chance": 10 },
```

File: data/mods/Magiclysm/worldgen/elf_workshop.json [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  },
+  {
+    "type": "overmap_special",
+    "id": "elf_workshop",
+    "overmaps": [
+      { "point": [ 0, 0, 0 ], "overmap": "elf_workshop_north" },
+      { "point": [ 0, 0, 1 ], "overmap": "elf_workshop_roof_north" }
+    ],
+    "locations": [ "wilderness" ],
+    "city_distance": [ 20, -5 ],
```
Comment (Procyonae) [inline data/mods/Magiclysm/worldgen/elf_workshop.json:76]:
```suggestion
    "city_distance": [ 20, -1 ],
```

File: data/mods/Magiclysm/worldgen/elf_workshop_roof.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
+[
+  {
+    "type": "mapgen",
+    "method": "json",
+    "om_terrain": [ "elf_workshop_roof" ],
+    "weight": 1000,
+    "object": {
```
Comment (Procyonae) [inline data/mods/Magiclysm/worldgen/elf_workshop_roof.json:7]:
```suggestion
    "om_terrain": [ "elf_workshop_roof" ],
    "object": {
```

**Comment by RDru** [comment]
This makes demon workshop and demon chitin too easy to obtain (imo)

**Pull request merged**